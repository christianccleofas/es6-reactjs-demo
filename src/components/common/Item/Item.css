.container {
    width: 20%; 
    display: inline-block; 
    padding: 2%;
    position: relative;
}

.floating-text-not-available {
    z-index: 100;
    position: absolute;
    top: 45%;
    left: 39%;
    font-size: 1.1em;
    font-weight: 600;
}

.item-button {
    background-color: #006241 !important;
}

.grp-btn {
    background-color: #006241 !important;
    box-shadow: none !important;
    border: solid 1px #006241;
    width: 36px !important;
    height: 36px !important;
    font-size: 1.3em;
    color: #fff;
}

.grp-btn-text {
    padding: 0px 10px;
    font-size: 1.1em;
    font-weight: 600;
    vertical-align: middle;
}

.test {
    background-color: #006241 !important;
    color: #fff !important;
}


.grp-btn.dcs {
    border-top-left-radius: 4px;
    border-bottom-left-radius: 4px;
}

.grp-btn.inc {
    border-top-right-radius: 4px;
    border-bottom-right-radius: 4px;
}

.grp-txt {
   box-shadow: none !important;
   border: solid 1px #006241;
   background-color: #fff;
   font-size: 1.5em;
}

.item-wrapper {
    cursor: pointer;
}

.item-wrapper-not-available {
    opacity: 0.5;
}

.item-display-wrapper {
    position: relative; 
    text-align: center;
}

.item-display-image {
    width: 100%;
}

.item-display-discount {
    position: absolute; 
    font-size: 1.3em; 
    font-weight: 600; 
    top: 10px; 
    right: 10px; 
    color: #fff;
}

.details {
    padding: 3%;
}

.item-name {
    display: inline-block;
    flex: 1;
    font-size: 1em; 
    font-weight: 600;
    color: #23423c;
    letter-spacing: -1;
}

.item-details-wrapper {
    flex: 1;
    text-align: right;
}

.item-details-amount {
    display: inline-block; 
    width: 50%; 
    font-size: 2.1em; 
    font-weight: 600; 
    text-align: right;
}

.item-details-amount-discounted {
    clear: both; 
    display:inline-block; 
    width: 50%; 
    text-align: right;
}

.item-details-amount-dsc {
    font-size: 1.3em; 
    font-weight: 600; 
    text-align: right;
    color: #AD0101;
}

.item-details-amount-n-dsc {
    font-size: 0.8em; 
    font-weight: 600; 
    text-align: right;
    text-decoration: line-through;
}

.item-details-addToCart {
    padding: 3%;
    border-top: solid 1px #eee;
    display: flex;
    align-items: center;
}

.add-to-cart {
    background: #006241;
    border: 0;
    color: white;
    font-size: small;
    font-weight: "400";
    padding: "5px 25px";
    box-shadow: 0 3px 5px 2px rgba(255, 105, 135, .3);
    flex: 1;
    text-transform: none;
}

.item-wrapper:hover {
    box-shadow: -1px 1px 11px 3px rgba(0,0,0,0.75);
    -webkit-box-shadow: -1px 1px 11px 3px rgba(0,0,0,0.75);
    -moz-box-shadow: -1px 1px 11px 3px rgba(0,0,0,0.75);
}