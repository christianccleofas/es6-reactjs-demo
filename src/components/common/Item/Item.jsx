import React from 'react';
import Paper from '@material-ui/core/Paper';
import AddIcon from '@material-ui/icons/Add';
import RemoveIcon from '@material-ui/icons/Remove';
import IconButton from '@material-ui/core/IconButton';
import "./Item.css";
import AddToCartButton from './AddToCartButton/AddToCartButton';
import _ from 'underscore';

const Item = ({ data, getOrderFromItem, orderList }) => {
	const [qty, setQty] = React.useState(0);

	const handleIncrease = () => {
		setQty(qty + 1);
	}

	const handleDecrease = () => {
		if (qty > 0) {
			setQty(qty - 1);
		}
	}

	const handleOrder = () => {
		const { name, price, discount } = data;
		let newOrderData = {
			name,
			price,
			qty,
			discount,
			discountedPrice() {
				return price - price * discount;
			},
			subTotal() {
				if (discount > 0) {
					return this.discountedPrice() * this.qty;
				} else {
					return price * this.qty;
				}
			},
		}
		getOrderFromItem(newOrderData);
	}

	React.useEffect(() => {
		//This useEffect function checks if the order is in the cart. if the user removes the order form the cart, it will reset the quantity back to 0.
		const isExisting = _.findWhere(orderList, {name: data.name});
		if(!isExisting) {
			setQty(0);
		}
	}, [orderList])

	return (
		<div className="container">
			{!data.isAvailable ? <span data-testid="span-not-available" className="floating-text-not-available">Not Available</span> : null}
			<Paper data-testid="paper" className={data.isAvailable ? "item-wrapper" : "item-wrapper-not-available"} elevation={3}>
				<div className="item-display-wrapper">
					<img className="item-display-image" src={`${process.env.PUBLIC_URL}/assets/images/${data.image}.webp`} alt="coffee-test" />
					{data.discount > 0 ? <div data-testid="div-display-discount" className="item-display-discount">{`-${data.discount*100}%`}</div> : null}
				</div>
				<div className="details" style={{ display: "flex", alignItems: "center" }}>
					<div className="item-name">
						{data.name}
					</div>
					<div className="item-details-wrapper">
						{data.discount && data.discount > 0 ?
							<div data-testid="div-amount-discount" className="item-details-amount-discounted">
								<div className="item-details-amount-n-dsc">
									{`$${data.price.toFixed(2)}`}
								</div>
								<div className="item-details-amount-dsc">
									{`$${(data.price - (data.price * data.discount)).toFixed(2)}`}
								</div>
							</div> :
							<div data-testid="div-amount" className="item-details-amount">{`$${data.price.toFixed(2)}`}</div>}
					</div>
				</div>
				<div className="item-details-addToCart">
					<div style={{ flex: '1' }}>
						<IconButton data-testid="button-subtract" className="test" size="small" onClick={handleDecrease}>
							<RemoveIcon fontSize="small" />
						</IconButton>
						<span date-testid="span-qty" className="grp-btn-text">{qty}</span>
						<IconButton data-testid="button-add" className="test" size="small" onClick={handleIncrease}>
							<AddIcon fontSize="small" />
						</IconButton>
					</div>
					<AddToCartButton data-testid="button-add-to-cart" className="add-to-cart" variant="contained" disabled={qty > 0 ? false : true} onClick={handleOrder}>
						Add to cart
					</AddToCartButton>
				</div>
			</Paper>
		</div>
	);
}

export default Item;