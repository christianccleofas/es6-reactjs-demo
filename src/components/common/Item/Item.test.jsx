import React from 'react';
import {shallow} from 'enzyme';
import Item from './Item';

const mocks = {
  data: {
    id: 0,
    name: "Test item",
    isAvailable: true,
    price: 3.66,
    image: "8e0b1749cfad49f085e3efff636aef58",
    discount: 0
  },
  getOrderFromItem: jest.fn(),
  orderList: [],
  qty: 0,
  useState: jest.fn(),
}

// jest.mock('react', () => ({
//   ...jest.requireActual('react'),
//   useState: x => [x, mocks.useState],
//   useEffect: f => f()
// }));
// const mockUseState = () => jest.spyOn(React, 'useState').mockReturnValueOnce([0, mocks.useState])

describe('<Item />: Component Testing', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('<Item />: Conditional render testing', () => {
    it('Should match snapshot.', () => {
      const wrapper = shallow(<Item data={mocks.data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
      expect(wrapper).toMatchSnapshot();
    });
    it('Should render with different elemets/styles if data.isAvailable is false.', () => {
      const data = {
        id: 0,
        name: "Test item",
        isAvailable: false,
        price: 3.66,
        image: "8e0b1749cfad49f085e3efff636aef58",
        discount: 0
      }
      const wrapper = shallow(<Item data={data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
      expect(wrapper.find('[data-testid="span-not-available"]').exists()).toBeTruthy();
      expect(wrapper.find('[data-testid="paper"]').props().className).toEqual('item-wrapper-not-available');
    });
    it('Should render with different elemets/styles if data.isAvailable is true.', () => {
      const data = {
        id: 0,
        name: "Test item",
        isAvailable: true,
        price: 3.66,
        image: "8e0b1749cfad49f085e3efff636aef58",
        discount: 0
      }
      const wrapper = shallow(<Item data={data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
      expect(wrapper.find('[data-testid="span-not-available"]').exists()).toBeFalsy();
      expect(wrapper.find('[data-testid="paper"]').props().className).toEqual('item-wrapper');
    });
    it('Should render with different elemets/styles if data.discount is not greater than 0.', () => {
      const data = {
        id: 0,
        name: "Test item",
        isAvailable: false,
        price: 3.66,
        image: "8e0b1749cfad49f085e3efff636aef58",
        discount: 0
      }
      const wrapper = shallow(<Item data={data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
      expect(wrapper.find('[data-testid="div-display-discount"]').exists()).toBeFalsy();
      expect(wrapper.find('[data-testid="div-amount"]').exists()).toBeTruthy();
      expect(wrapper.find('[data-testid="div-amount-discount"]').exists()).toBeFalsy();
    });
    it('Should render with different elemets/styles if data.discount is greater than 0.', () => {
      const data = {
        id: 0,
        name: "Test item",
        isAvailable: false,
        price: 3.66,
        image: "8e0b1749cfad49f085e3efff636aef58",
        discount: 5
      }
      const wrapper = shallow(<Item data={data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
      expect(wrapper.find('[data-testid="div-display-discount"]').exists()).toBeTruthy();
      expect(wrapper.find('[data-testid="div-amount"]').exists()).toBeFalsy();
      expect(wrapper.find('[data-testid="div-amount-discount"]').exists()).toBeTruthy();
    });
  });

  describe('<Item />: Event handler testing without mocked states', () => {
    test('onClick: button-add should trigger handleIncrease', () => {
      const wrapper = shallow(<Item data={mocks.data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
      expect(wrapper.find('[date-testid="span-qty"]').text()).toEqual('0');
      wrapper.find('[data-testid="button-add"]').simulate('click'); // +1
      expect(wrapper.find('[date-testid="span-qty"]').text()).toEqual('1');
    });
    test('onClick: button-subtract should trigger handleDecrease', () => {
      const wrapper = shallow(<Item data={mocks.data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
      expect(wrapper.find('[date-testid="span-qty"]').text()).toEqual('0');
      wrapper.find('[data-testid="button-add"]').simulate('click'); // +1
      wrapper.find('[data-testid="button-subtract"]').simulate('click'); // -1
      expect(wrapper.find('[date-testid="span-qty"]').text()).toEqual('0');
    });
  });

  // describe('<Item />: Event handler testing', () => {
    // beforeEach(() => {
    //   jest.clearAllMocks()
    // })
    // test('onClick: button-add should trigger handleIncrease', () => {
    //   const wrapper = shallow(<Item data={mocks.data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
    //   wrapper.find('[data-testid="button-add"]').simulate('click');
    //   expect(mocks.useState).toHaveBeenCalledWith(1);
    // });
    // test('onClick: button-subtract should trigger handleDecrease', () => {
    //   const wrapper = shallow(<Item data={mocks.data} getOrderFromItem={mocks.getOrderFromItem} orderList={mocks.orderList} />);
    //   wrapper.find('[data-testid="button-add"]').simulate('click');
    //   wrapper.find('[data-testid="button-subtract"]').simulate('click');
    //   console.log(mocks.useState.mock.calls)
    //   expect(mocks.useState.mock.calls[1].length).toBe(0);
    // });
  // });
});
