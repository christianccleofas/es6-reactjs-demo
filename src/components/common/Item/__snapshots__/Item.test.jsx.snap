// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item />: Component Testing <Item />: Conditional render testing Should match snapshot. 1`] = `
<div
  className="container"
>
  <WithStyles(ForwardRef(Paper))
    className="item-wrapper"
    data-testid="paper"
    elevation={3}
  >
    <div
      className="item-display-wrapper"
    >
      <img
        alt="coffee-test"
        className="item-display-image"
        src="/assets/images/8e0b1749cfad49f085e3efff636aef58.webp"
      />
    </div>
    <div
      className="details"
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <div
        className="item-name"
      >
        Test item
      </div>
      <div
        className="item-details-wrapper"
      >
        <div
          className="item-details-amount"
          data-testid="div-amount"
        >
          $3.66
        </div>
      </div>
    </div>
    <div
      className="item-details-addToCart"
    >
      <div
        style={
          Object {
            "flex": "1",
          }
        }
      >
        <WithStyles(ForwardRef(IconButton))
          className="test"
          data-testid="button-subtract"
          onClick={[Function]}
          size="small"
        >
          <Memo(ForwardRef(RemoveIcon))
            fontSize="small"
          />
        </WithStyles(ForwardRef(IconButton))>
        <span
          className="grp-btn-text"
          date-testid="span-qty"
        >
          0
        </span>
        <WithStyles(ForwardRef(IconButton))
          className="test"
          data-testid="button-add"
          onClick={[Function]}
          size="small"
        >
          <Memo(ForwardRef(AddIcon))
            fontSize="small"
          />
        </WithStyles(ForwardRef(IconButton))>
      </div>
      <WithStyles(WithStyles(ForwardRef(Button)))
        className="add-to-cart"
        disabled={true}
        onClick={[Function]}
        variant="contained"
      >
        Add to cart
      </WithStyles(WithStyles(ForwardRef(Button)))>
    </div>
  </WithStyles(ForwardRef(Paper))>
</div>
`;
